RTI-Connext-DDS external dependency Quality Declaration (#402)
ros2/rmw_connext: Connext_DDS_QUALITY_DECLARATION.md

This Quality Declaration claims that the external dependency `RTI-Connext-DDS` is in the Quality Level 4 category.

Below are the rationales, notes, and caveats for this claim, organized by each requirement listed in the [Package Requirements for Quality Level 4 in REP-2004](https://www.ros.org/reps/rep-2004.html).

## Summary
`RTI-Connext-DDS` meets most of the standards defined for Quality Level 1, including broad documentation and testing of its features and API.

As this package is not open-source, many policies related to quality control are not openly documented. For example, it is difficult to find information relating to its versioning policy. For ROS 2 distributions, packages may need to rely on strict versions, like `5.3.1`, to ensure API/ABI stability. 

There are no clear public reports of code coverage, so it is not clear how much of their code is tested. However, they do describe their approach to testing in a [blog entry](https://www.rti.com/blog/software-testing-at-rti). 

> There is a list of [known academic projects](https://community.rti.com/projects) using RTI Connext DDS and also they have published a [white paper](https://www.rti.com/whitepapers/how-to-achieve-production-grade-deployment-with-ros-2-and-rti-connext-dds) with rationale about how to implement production-grade deployments using their software in conjunction with ROS 2.

Reviewer: Since QDs attempt to describe the suitability of a package for production use, it might be better to document connext's uses in industry production systems rather than academic ones. I think you should keep the ros 2 whitepaper, but instead of academic uses maybe describe and document that Connext is used in aerospace, aviation, automotive, healthcare, oil & gas, etc.

> In terms of ROS 2 package metrics this library is considered to be Quality Level 4. Adding unit testing for the functions used in ROS 2 packages, coverage statistics and version pinning will be needed to achieve Quality Level 1.

Reviewer: It should be made clear who needs to do what to increase the quality level. RTI doesn't really have to add unit testing, or pin versions. Instead, they just need to be more transparent about their code coverage, API coverage, versioning policy, etc in order for consumers of the package to better understand how to depend on it.

### Public API Declaration [1.iii]

The public API is declared in their [documentation page](https://community.rti.com/documentation/rti-connext-dds-531) with sections for [C API](https://community.rti.com/static/documentation/connext-dds/5.3.1/doc/api/connext_dds/api_c/index.html), [C++ Modern API](https://community.rti.com/static/documentation/connext-dds/5.3.1/doc/api/connext_dds/api_cpp2/index.html), [C ++ Traditional API](https://community.rti.com/static/documentation/connext-dds/5.3.1/doc/api/connext_dds/api_cpp/index.html) and other programming languages.

### API Stability Policy [1.iv]/[1.vi]

There is no declared policy for API stability and it's unclear if breaking changes may be introduced during minor or patch versions. Packages that depend on `RTI-Connext-DDS` may need to limit the versions they rely on to ensure API stability.

### ABI Stability Policy [1.v]/[1.vi]

There is no declared policy for ABI stability and it's unclear if breaking changes may be introduced during minor or patch versions. Packages that depend on `RTI-Connext-DDS` may need to limit the versions they rely on to ensure API stability.

## Change Control Process [2]

> Change control process is not publicly documented (sections [2.i - 2.v]).

Reviewer: ```suggestion
The change control process of `RTI-Connext-DDS` is not publicly documented (sections [2.i - 2.v]).
```

## Documentation [3]

### Feature Documentation [3.i]

> The complete software provides extensive and high quality documentation hosted at [main documentation page](https://community.rti.com/documentation/rti-connext-dds-531).

Reviewer: ```suggestion
`RTI-Connext-DDS` provides extensive and high quality documentation hosted on their [main documentation page](https://community.rti.com/documentation/rti-connext-dds-531).
```

### Linters and Static Analysis [4.v]

There is no public information available about this software's linting or static analysis.

## Dependencies [5]

### Direct/Optional Runtime ROS Dependencies [5.i]/[5.ii]

`RTI-Connext-DDS` has no direct/optional runtime ROS dependencies.

### Direct Runtime non-ROS Dependency [5.iii]

`RTI-Connext-DDS` depends on the external library `openssl1.1.1`. This library comes compiled and bundled with the RTI software. It is expected that this library is tested as a part of the `RTI-Connext-DDS` testing tools environment.

## Platform Support [6]

The `RTI-Connext-DDS` provides support for Linux, MacOs and Windows platforms, as it can be seen in their [supported platforms](https://community.rti.com/static/documentation/connext-dds/5.3.1/doc/manuals/connext_dds/html_files/RTI_ConnextDDS_CoreLibraries_ReleaseNotes/index.htm#ReleaseNotes/System_Requirements.htm#Table_SupportedPlatforms) landing page.  However, there is no declared support for Ubuntu Focal nor MacOs 10.14.

## Security [7]

###  Vulnerability Disclosure Policy [7.i]

> `RTI-Connext-DDS` does not have a Vulnerability Disclosure Policy.

Reviewer: This still surprises me, but I can't find anything myself. Or even reports of security vulnerability disclosures.

The main concerns related to this software are related to the fact that the development of it is not complete in an open-source fashion, so it is unknown the policies they have to develop their code and if they have any policies related to versioning. For the case of ROS2, as the libraries are fixed to the `5.3.1` there should not be any considering versioning and API/ABI stability. 

Without reports of coverage, it is not clear if the software is completely tested. However, according to a [blog entry](https://www.rti.com/blog/software-testing-at-rti)  from RTI, the company takes testing very seriously. 

There is a list of [known academic projects](https://community.rti.com/projects) using RTI Connext DDS and also they have published a [white paper](https://www.rti.com/whitepapers/how-to-achieve-production-grade-deployment-with-ros-2-and-rti-connext-dds) with rationale about how to implement production-grade deployments using their software in conjunction with ROS2.

> In terms of ROS2 package metrics this library is considered to be Quality Level 4. Adding unit testing for the functions used in ROS2 packages, coverage statistics and version pinning will be needed to achieve Quality Level 1.

Author: I will change the paragraph. I did think they probably had a lot of the missing parts already covered, I also think it's important to encourage them to state their practices in terms of REP-2004 to so the whole community can aim for higher levels of quality in their packages.

# Comparison with ROS packages quality standards

## Version Policy [1]

> 

Author: Done
